msal/tests/Test.MSAL.NET.Integration/UnifiedCacheTests.cs

```diff
@@ -0,0 +1,491 @@
+                AcquireTokenAsync(adalContext, AdalResource2, ClientId,
+                new adal::Microsoft.IdentityModel.Clients.ActiveDirectory.UserPasswordCredential(user.Upn, securePassword));
+            ValidateAdalAuthResult();
+            
+            // simulate adalV3 token cache state by setting client info in adal cache entities to null 
+            // and clearing msal cache
+            UpdateAdalCacheSetClientInfoToNull();
+            ClearMsalCache();
+            AssertMsalCacheIsEmpty();
+
+            // make sure that adal v3 RT is visible for Msal
+            var msalAccounts = await msalPublicClient.GetAccountsAsync();
+            Assert.AreEqual(1, msalAccounts.Count());
+            var account = msalAccounts.First();
+            Assert.AreEqual(user.Upn, account.Username);
+            Assert.IsNull(account.HomeAccountId);
+            Assert.IsNotNull(account.Environment);
+
+            // make sure that adal v3 RT is usable by Msal
+            msalAuthResult = await msalPublicClient.AcquireTokenSilentAsync(MsalScopes, account);
+            ValidateMsalAuthResult();
+
+            // make sure Msal remove account api remove corresponding cache entities in all formats  
+            msalAccounts = await msalPublicClient.GetAccountsAsync();
+            Assert.AreEqual(1, msalAccounts.Count());
+            account = msalAccounts.First();
+
+            await msalPublicClient.RemoveAsync(account);
+
+            AssertAdalCacheIsEmpty();
+            AssertMsalCacheIsEmpty();
+        }
+
+        [TestMethod]
+        [TestCategory("UnifiedCache_IntegrationTests")]
+        public async Task UnifiedCache_Adalv3ToAdalV4ToMsal2MigrationIntegrationTest()
+        {
+            // acquire adal tokens using adalV4
+            adalAuthResult = await adal::Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContextIntegratedAuthExtensions.
+                AcquireTokenAsync(adalContext, AdalResource1, ClientId,
+                new adal::Microsoft.IdentityModel.Clients.ActiveDirectory.UserPasswordCredential(user.Upn, securePassword));
+            ValidateAdalAuthResult();
+
+            // simulate adalV3 token cache state by setting client info in adal cache entities to null 
+            // and clearing msal cache
+            UpdateAdalCacheSetClientInfoToNull();
+            ClearMsalCache();
+            AssertMsalCacheIsEmpty();
+
+            // Migration to AdalV4 - acquire adal tokens using adalV4
+            adalAuthResult = await adal::Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContextIntegratedAuthExtensions.
+                AcquireTokenAsync(adalContext, AdalResource2, ClientId,
+                new adal::Microsoft.IdentityModel.Clients.ActiveDirectory.UserPasswordCredential(user.Upn, securePassword));
+            ValidateAdalAuthResult();
```

Comment:
suggestion before this add a silent call using adal v4 to the resource you already requested an AT for... that way you will be able to validate that v4 works even with-out user_info... or perhaps that it doesn't work

Reply:
good point will add